text/0000-format-args-implicit-identifiers.md

```diff
@@ -0,0 +1,340 @@
+
+The following code would be the equivalent way to produce a new string combining a `greeting` and a `person` in a variety of languages:
+
+    // Rust
+    format!("{} {}", greeting, person)                                // positional form,
+    format!("{greeting} {person}", greeting=greeting, person=person)  // or named form
+
+    # Python 3
+    f"{greeting} {person}"
+
+    // Javascript
+    `${greeting} ${person}`
+
+    // C# / VB
+    $"{greeting} {person}"
+
+    // Swift
+    "\(greeting) \(person)"
+
+    // Ruby
+    "#{greeting} #{person}"
+
+    // Scala
+    s"$greeting $person"
+    
+    // PHP
+    "$greeting $person"
+
+It is the RFC author's experience that these interpolating mechanisms read easily from left-to-right and it is clear where each variable is being substituted into the format string.
+
+In the Rust formatting macros as illustrated above, the positional form suffers the drawback of not reading strictly from left to right; the reader of the code must refer back-and-forth between the format string and the argument list to determine where each variable will be subsituted. The named form avoids this drawback at the cost of much longer code.
+
+Implementing implicit named arguments in the fashion suggested in this RFC would eliminate the drawbacks of each of the Rust forms and permit new syntax much closer to the other languages:
+
+    // Rust - implicit named arguments
+    format!("{greeting} {person}")
+
+It should be noted, however, that other languages' string interpolation mechanisms allow substitution of a wide variety of expressions beyond the simple identifier case that this RFC is focussed on.
+
+Please see the discusison on [interpolation](#interpolation) as an alternative to this RFC.
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+## Should implicit named arguments be accepted for formatting parameters?
+
+Some of the formatting traits can accept additional formatting parameters to control how the argument is displayed. For example, the precision with which to display a floating-point number:
+
+    println!("{:.5}", x);  // print x to 5 decimal places
+
+It is also possible for the precision to refer to either positional or named arguments using "dollar syntax":
+
+    println!("{:.1$}", x, 5);
+    println!("{:.prec$}", x, prec=5);
+
+As a result of this RFC, formatting parameters could potentially also accept implicit named arguments:
+
+    println!("{x:.precision$}");
+
+The RFC author believes Rust users familiar with implicit named arguments may expect the above to compile (as long as `x` and `precision` were valid identifiers in the scope in question). However, feedback is requested during this RFC process as to whether the this should be indeed become acceptable as part of the RFC.
```

Comment:
I would very much like implicit arguments to work anywhere explicitly provided named arguments currently work, for consistency.

One interesting interaction that seems worth noting: because those named arguments appear right next to the `.`, if we want to allow `dotted.paths` in the future, we could end up with something like `{x:self.width$.self.precision$}`. I don't know how easily we could parse that, and it certainly confuses my mental parser. Could you discuss this a bit, and suggest possible future solutions *if* we decide to extend this to dotted paths? (I agree with your rationale of not doing dotted paths in *this* RFC, but it does seem like the most obvious extension, and worth discussing future-proofing for.)

Reply:
It's a great question about `dotted.paths`, and one that I've been thinking hard about. I agree with you that `{x:self.width$.self.precision$}` also confuses me to read. This indicates to me that we probably don't want to end up with implicit dotted paths in formatting parameters point.

Reply:
Even in just making the step to permit `dotted.paths`, I think that using the interpolation syntax suggested by @kennytm, i.e. `{(expr})`, is likely to be the best option.

My reasoning: It offers improved readability in the basic case for complex expressions, and seems to me to also help with this problem you highlight in the formatting parameters case. For example, we could have
       
       {x:(self.width).(self.precision)}        // seems okay to me
       {x:(self.width)$.(self.precision)$}      // or an alternative if we decide to keep the postfix dollars

Both of these read okay to me.

A good argument can be made that if we added interpolation we might aswell go all the way and allow permit any expression for interpolation in `{(expr)}`. But this doesn't have to be the case - we could, for example, test the waters with just `{(dotted.paths)}`.

While it's tempting to wish for `{dotted.paths}` without the brackets, I suspect in the long run we'll perfer `{(expr)}` instead of `{expr}` if we want to start adding any interpolation.

Reply:
A problem with "dollar syntax" is that it's horribly hard to find if you don't know what you're looking for, and confusing if you don't know it already.

The solution of just using `(expr)` (assuming it's not ambiguous anywhere; be careful around the pad character which can be anything, including `}`(!)) makes it a lot more obvious what is going on, I think.

This actually would put some weight on always using the `(expr)` form, as `{(x):(width).(precision)}` is more internally consistent than `{x:width$.precision$}`.

That said, I agree that if we allow `{x}`, the rule should be just "if named argument isn't provided, look for it in scope", no matter the source of that named argument.

Reply:
> This actually would put some weight on always using the (expr) form, as {(x):(width).(precision)} is more internally consistent than {x:width$.precision$}.

I do agree with this statement. Though `{x:width$.precision$}` is what we _already_ have for named arguments, and I'd much rather avoid introducing new syntax when the existing works for the proposal at hand.

Seems most (all?) of us are arguing the `{(expr)}` form is better for forwards compatibility with interpolation. However, if we added `{(expr)}` interpolation in the future, the `{(ident)}` special case will just as equally be backwards compatible with `{ident}`. So I would argue to avoid introducing new `{(expr)}` syntax now if we're not sure we will want to add interpolation to these macros in the future.

Reply:
@davidhewitt If we considering the possibility that future syntax extensions might potentially change the dollar syntax, I like the idea of just assuming that we'll need parentheses to put `dotted.names` in the current syntax, and then considering potential alternative syntaxes for width and precision that parse unambiguously without requiring parentheses. Would you consider listing that as an alternative?

Reply:
Yes, great idea, have now done so.